mono/metadata/cominterop.c

```diff
@@ -3699,32 +3669,46 @@ mono_cominterop_get_com_interface (MonoObject *object, MonoClass *ic, MonoError
 			return NULL;
 		klass = mono_object_class (object);
 		if (!mono_class_is_transparent_proxy (klass)) {
+			if (icall)
+				g_assert_not_reached ();
```

Comment:
I think it's better to `g_assert (mono_class_is_transparent_proxy (klass))` in `..._internal` and to do the dynamic check in `mono_cominterop_get_com_interface`.